sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/RequestResponseIO.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.PValue;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap;
+
+/**
+ * {@link PTransform} for reading from and writing to Web APIs.
+ *
+ * <h2>Basic Usage</h2>
+ *
+ * {@link RequestResponseIO} minimally requires implementing the {@link Caller} interface:
+ *
+ * <pre>{@code class MyCaller implements Caller<SomeRequest, SomeResponse> {
+ *    public SomeResponse call(SomeRequest request) throws UserCodeExecutionException {
+ *      // calls the API submitting SomeRequest payload and returning SomeResponse
+ *    }
+ * }}</pre>
+ *
+ * <p>Then provide {@link RequestResponseIO}'s {@link #create} method your {@link Caller}
+ * implementation.
+ *
+ * <pre>{@code  PCollection<SomeRequest> requests = ...
+ *  Result result = requests.apply(RequestResponseIO.create(new MyCaller()));
+ *  result.getResponses().apply( ... );
+ *  result.getFailures().apply( ... );
+ * }</pre>
+ */
+public class RequestResponseIO<RequestT, ResponseT>
+    extends PTransform<PCollection<RequestT>, Result<ResponseT>> {
+
+  private static final TupleTag<ApiIOError> FAILURE_TAG = new TupleTag<ApiIOError>() {};
+
+  // TODO(damondouglas): remove when utilized.
+  @SuppressWarnings({"unused"})
+  private final Configuration<RequestT, ResponseT> configuration;
+
+  private RequestResponseIO(Configuration<RequestT, ResponseT> configuration) {
+    this.configuration = configuration;
+  }
+
+  public static <RequestT, ResponseT> RequestResponseIO<RequestT, ResponseT> create(
+      Caller<RequestT, ResponseT> caller) {
+    return new RequestResponseIO<>(
+        Configuration.<RequestT, ResponseT>builder().setCaller(caller).build());
+  }
+
+  /** Configuration details for {@link RequestResponseIO}. */
+  @AutoValue
+  abstract static class Configuration<RequestT, ResponseT> {
+
+    static <RequestT, ResponseT> Builder<RequestT, ResponseT> builder() {
+      return new AutoValue_RequestResponseIO_Configuration.Builder<>();
+    }
+
+    /**
+     * The {@link Caller} that interfaces user custom code to process a {@link RequestT} into a
+     * {@link ResponseT}.
+     */
+    abstract Caller<RequestT, ResponseT> getCaller();
+
+    abstract Builder<RequestT, ResponseT> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<RequestT, ResponseT> {
+
+      abstract Builder<RequestT, ResponseT> setCaller(Caller<RequestT, ResponseT> value);
+
+      abstract Configuration<RequestT, ResponseT> build();
+    }
+  }
+
+  @Override
+  public Result<ResponseT> expand(PCollection<RequestT> input) {
+    // TODO(damondouglas; https://github.com/apache/beam/issues?q=is%3Aissue+is%3Aopen+%5BRRIO%5D):
```

Comment:
This will be the interesting part :)

Reply:
Thank you. I will go ahead and resolve this comment to keep the PR uncluttered. Thank you, again, for reviewing.